cmake/rgw: apply -DCLS_CLIENT_HIDE_IOCTX to rgw_common too

cbodley opened the pull request:
this CLS_CLIENT_HIDE_IOCTX define from https://github.com/ceph/ceph/pull/28569 wasn't applying to the rgw_common target, so several blocking cls calls there went unnoticed

add/use ObjectOperation overloads for those and submit them with rgw_rados_operate()

finally, move the CLS_CLIENT_HIDE_IOCTX define to the rgw_common target so it applies to all rgw targets


<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

cbodley committed:
cls/rgw: enable async cls_rgw_reshard_get/list

get and list take ObjectReadOperation so they can be used with aio_operate()

the RGWReshard::get() and list() functions now take optional_yield and
call rgw_rados_operate()

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
cmake/rgw: apply -DCLS_CLIENT_HIDE_IOCTX to rgw_common too

move the definition of CLS_CLIENT_HIDE_IOCTX from rgw_a to rgw_common.
the definition is PUBLIC so doesn't need to be repeated for other
dependent targets

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
cls/otp: get_current_time() takes ObjectReadOperation

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
cls/otp: split check() into separate functions

each taking the corresponding ObjectOperation so they can be used in
asynchronous operations. call them sequentially in
RGWSI_Cls::MFA::check_mfa() where rgw can use rgw_rados_operate() to
suspend the calling coroutine while waiting

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/notifications: call ObjectOperation overload of get_topic_stats()

avoid the blocking overload taking IoCtx because these are called by
coroutines

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/notifications: call ObjectOperation overload of list_reservations()

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/gc: use ObjectOperation overload of cls_version_read()

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
cls/rgw: usage_log_read() takes ObjectReadOperation

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley commented on src/rgw/services/svc_cls.cc:
```diff
@@ -51,9 +51,28 @@ int RGWSI_Cls::MFA::check_mfa(const DoutPrefixProvider *dpp, const rgw_user& use
   }
 
-  rados::cls::otp::otp_check_t result;
+#define TOKEN_LEN 16
```
would you prefer `constexpr size_t TOKEN_LEN = 16` or `gen_rand_alphanumeric(cct, 16)`?

adamemerson replied:
The first, it's just sticking defines in the middle of function bodies as if they were local variables that really bugs me.

adamemerson commented on src/rgw/services/svc_cls.cc:
```diff
@@ -51,7 +51,7 @@ int RGWSI_Cls::MFA::check_mfa(const DoutPrefixProvider *dpp, const rgw_user& use
   }
 
-#define TOKEN_LEN 16
+  constexpr size_t TOKEN_LEN = 16;
```
:+1:

cbodley committed:
rgw: remove unused lock methods from RGWMetadataLog

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: remove unused RGWSI_Cls::Lock subservice

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: remove unused RGWOp_MDLog_Lock/Unlock

these /admin/log apis were from the ancient sync agent that predated
multisite v3 and have gone unused since

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: RGWObjectExpirer uses ObjectOperation overloads for cls_lock

Signed-off-by: Casey Bodley <cbodley@redhat.com>

adamemerson commented:
jenkins test api

cbodley committed:
rgw/reshard: RGWBucketReshardLock takes optional_yield

avoid blocking on cls_lock when called from a coroutine

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw: RGWAsyncLockSystemObj uses ObjectOperation overloads of cls_lock

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/gc: use ObjectOperation overloads of cls_lock

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/rados: RadosRestoreSerializer uses ObjectOperation

avoid calling synchronous overload of lock_exclusive()

Signed-off-by: Casey Bodley <cbodley@redhat.com>

cbodley committed:
rgw/dedup: remove unused function calling cls_refcount_read

the IoCtx overload is hidden so this fails to compile

Signed-off-by: Casey Bodley <cbodley@redhat.com>
